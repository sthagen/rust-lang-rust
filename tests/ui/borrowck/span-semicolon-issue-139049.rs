// Make sure the generated suggestion suggest editing the user
// code instead of the std macro implementation

//@ run-rustfix

#![allow(dead_code)]

use std::fmt::{self, Display};

struct Mutex;

impl Mutex {
    fn lock(&self) -> MutexGuard<'_> {
        MutexGuard(self)
    }
}

struct MutexGuard<'a>(&'a Mutex);

impl<'a> Drop for MutexGuard<'a> {
    fn drop(&mut self) {}
}

struct Out;

impl Out {
    fn write_fmt(&self, _args: fmt::Arguments) {}
}

impl<'a> Display for MutexGuard<'a> {
    fn fmt(&self, _formatter: &mut fmt::Formatter) -> fmt::Result {
        Ok(())
    }
}

fn main() {
    let _write = {
        let mutex = Mutex;
        write!(Out, "{}", mutex.lock())
        //~^ ERROR `mutex` does not live long enough
        //~| SUGGESTION ;
    };

    let _write = {
        use std::io::Write as _;

        let mutex = Mutex;
        write!(std::io::stdout(), "{}", mutex.lock())
        //~^ ERROR `mutex` does not live long enough
        //~| SUGGESTION let x
    };
}
