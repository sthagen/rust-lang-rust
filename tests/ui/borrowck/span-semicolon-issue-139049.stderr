error[E0597]: `mutex` does not live long enough
  --> $DIR/span-semicolon-issue-139049.rs:39:27
   |
LL |         let mutex = Mutex;
   |             ----- binding `mutex` declared here
LL |         write!(Out, "{}", mutex.lock())
   |                           ^^^^^-------
   |                           |
   |                           borrowed value does not live long enough
   |                           a temporary with access to the borrow is created here ...
...
LL |     };
   |     -- ... and the borrow might be used here, when that temporary is dropped and runs the `Drop` code for type `MutexGuard`
   |     |
   |     `mutex` dropped here while still borrowed
   |
help: consider adding semicolon after the expression so its temporaries are dropped sooner, before the local variables declared by the block are dropped
   |
LL |         write!(Out, "{}", mutex.lock());
   |                                        +

error[E0597]: `mutex` does not live long enough
  --> $DIR/span-semicolon-issue-139049.rs:48:41
   |
LL |         let mutex = Mutex;
   |             ----- binding `mutex` declared here
LL |         write!(std::io::stdout(), "{}", mutex.lock())
   |                                         ^^^^^-------
   |                                         |
   |                                         borrowed value does not live long enough
   |                                         a temporary with access to the borrow is created here ...
...
LL |     };
   |     -- ... and the borrow might be used here, when that temporary is dropped and runs the `Drop` code for type `MutexGuard`
   |     |
   |     `mutex` dropped here while still borrowed
   |
   = note: the temporary is part of an expression at the end of a block;
           consider forcing this temporary to be dropped sooner, before the block's local variables are dropped
help: for example, you could save the expression's value in a new local variable `x` and then make `x` be the expression at the end of the block
   |
LL |         let x = write!(std::io::stdout(), "{}", mutex.lock()); x
   |         +++++++                                              +++

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0597`.
